File: C#-NET-Stuff/JavaExamples/SampleJavaApp/src/main/java/org/example/model/SavingsAccount.java
package org.example.model;

import org.example.exception.InvalidDepositAmountException;
import org.example.exception.InvalidWithdrawAmountException;
import org.example.exception.UnableToAssignAccountNumberException;

public class SavingsAccount extends Account{

    public static final double SAVINGS_DEFAULT_BALANCE = 0;
    public static final double SAVINGS_DEFAULT_APR     = .01;

    private double annualInterestRate;

    public SavingsAccount(String acctOwner, double balance, double annualInterestRate) throws UnableToAssignAccountNumberException {
        super(acctOwner, balance);
        this.annualInterestRate = annualInterestRate;
    }
    public SavingsAccount(String acctOwner, double balance) throws UnableToAssignAccountNumberException {
        super(acctOwner, balance);
        this.annualInterestRate = SAVINGS_DEFAULT_APR;
    }
    public SavingsAccount(String acctOwner) throws UnableToAssignAccountNumberException {
        super(acctOwner, SAVINGS_DEFAULT_BALANCE);
        this.annualInterestRate = SAVINGS_DEFAULT_APR;
    }

    public double getAnnualInterestRate() {
        return annualInterestRate;
    }

    public void setAnnualInterestRate(double annualInterestRate) {
        this.annualInterestRate = annualInterestRate;
    }

    @Override
    public double withdraw(double amount) {
        if (amount > balance || amount < 0) {
            throw new InvalidWithdrawAmountException("Withdrawal amount ("+ amount + ") exceeds balance (" + balance + ") or is less than 0");
        }
        return balance -= amount;
    }

    @Override
    public double deposit(double amount) {
        if (amount < 0) {
            throw new InvalidDepositAmountException("Withdrawal amount (" + amount + ") cannot be less than 0");
        }
        return balance += amount;
    }

    public double applyMonthlyInterest() {
        return balance += balance * (annualInterestRate / 12);
    }

    @Override
    public String toString() {
        return super.toString() + "; \nSavingsAccount{" +
                "annualInterestRate=" + annualInterestRate +
                '}';
    }
}

File: C#-NET-Stuff/example-code/SampleCSHarp-VS2022-Net6/SampleCSharpAppTests/Model/SavingsAccount.cs
using SampleCSharpAppTests.Exception;

namespace SampleCSharpAppTests.Model
{
    public class SavingsAccount : Account
    {
        public const double SavingsDefaultBalance = 0;
        public const double SavingsDefaultApr = 0.01;

        private double _annualInterestRate;

        public SavingsAccount(string acctOwner, double balance, double annualInterestRate)
            : base(acctOwner, balance)
        {
            this._annualInterestRate = annualInterestRate;
        }

        public SavingsAccount(string acctOwner, double balance)
            : base(acctOwner, balance)
        {
            this._annualInterestRate = SavingsDefaultApr;
        }

        public SavingsAccount(string acctOwner)
            : base(acctOwner, SavingsDefaultBalance)
        {
            this._annualInterestRate = SavingsDefaultApr;
        }

        public double GetAnnualInterestRate()
        {
            return _annualInterestRate;
        }

        public void SetAnnualInterestRate(double annualInterestRate)
        {
            this._annualInterestRate = annualInterestRate;
        }

        public override double Withdraw(double amount)
        {
            if (amount > Balance || amount < 0)
            {
                throw new InvalidWithdrawAmountException("Withdrawal amount (" + amount + ") exceeds balance (" + Balance + ") or is less than 0");
            }
            return Balance -= amount;
        }

        public override double Deposit(double amount)
        {
            if (amount < 0)
            {
                throw new InvalidDepositAmountException("Withdrawal amount (" + amount + ") cannot be less than 0");
            }
            return Balance += amount;
        }

        public double ApplyMonthlyInterest()
        {
            return Balance += Balance * (_annualInterestRate / 12);
        }

        public override string ToString()
        {
            return base.ToString() + "; \nSavingsAccount{" +
                   "annualInterestRate=" + _annualInterestRate +
                   '}';
        }

    }
}

